### PR TITLE
Add EFI bootchooser backend for x86

### PR DESCRIPTION
**Pull request by ejoerns**
This series adds support for handling EFI-based booting on x86 machines.

It calls and parses output of the `efibootmgr` tool to modify and read the EFI variables required for boot order modification.

### CODE REVIEW

File: src/bootchooser.c [hunk 1]
```diff
@@ -614,6 +615,468 @@ static gboolean uboot_set_primary(RaucSlot *slot, GError **error) {
...
+		g_set_error(
+				error,
+				R_BOOTCHOOSER_ERROR,
+				R_BOOTCHOOSER_ERROR_FAILED,
+				"Regex matching failed!");
+		res = FALSE;
+		goto out;
+	}
+
+	while (g_match_info_matches (match)) {
```
Comment (jluebbe) [inline src/bootchooser.c]: Avoid spaces after the function name also below.

File: src/bootchooser.c [hunk 2]
```diff
@@ -614,6 +615,468 @@ static gboolean uboot_set_primary(RaucSlot *slot, GError **error) {
...
+				R_BOOTCHOOSER_ERROR,
+				R_BOOTCHOOSER_ERROR_FAILED,
+				"unable to obtain boot order!");
+		res = FALSE;
+		goto out;
+	}
+
+	bootnumorder = g_strsplit(g_match_info_fetch(match, 1), ",", 0);
+
+	/* Iterate over boot counter list in boot order */
```
Comment (jluebbe) [inline src/bootchooser.c]: This doesn't iterate over boot counters, but boot entries.

File: src/bootchooser.c [hunk 3]
```diff
@@ -614,6 +615,468 @@ static gboolean uboot_set_primary(RaucSlot *slot, GError **error) {
...
+		if (g_strcmp0(bootnum, ptr->num) == 0) {
+			found_entry = ptr;
+			break;
+		}
+	}
+
+	return found_entry;
+}
+
+static gboolean efi_bootorder_get(GList **bootorder_entries, GList **all_entries, efi_bootentry **bootnext, GError **error) {
```
Comment (jluebbe) [inline src/bootchooser.c]: This needs a comment to explain the arguments, especially bootorder_entries and all_entries.

File: src/bootchooser.c [hunk 4]
```diff
@@ -614,6 +615,468 @@ static gboolean uboot_set_primary(RaucSlot *slot, GError **error) {
...
+
+	bootorder = g_ptr_array_sized_new(g_list_length(entries));
+	/* Construct bootorder string out of boot entry list */
+	for (GList *entry = entries; entry != NULL; entry = entry->next) {
+		efi_bootentry *efi = entry->data;
+		g_ptr_array_add(bootorder, efi->num);
+	}
+	g_ptr_array_add(bootorder, NULL);
+
+	order = g_strjoinv(",", (gchar**) g_ptr_array_free(bootorder, FALSE));
```
Comment (jluebbe) [inline src/bootchooser.c]: This needs a comment why we don't need to free the individual strings.

File: src/bootchooser.c [hunk 5]
```diff
@@ -614,6 +615,468 @@ static gboolean uboot_set_primary(RaucSlot *slot, GError **error) {
...
+
+	res = efi_bootorder_set(order, NULL);
+	if (!res) {
+		g_propagate_prefixed_error(error, ierror, "Modifying bootorder failed: ");
+		goto out;
+	}
+
+	res = TRUE;
+out:
+	return res;
```
Comment (jluebbe) [inline src/bootchooser.c]: We need to free the order string.

**Comment by ejoerns** [comment]
I've upated the patch series with some fixes resulting from further tests, changes as requested by review and added documentation.

**Pull request merged**